src/Core/MailTemplate/MailLayoutFolderCatalog.php

```diff
@@ -0,0 +1,246 @@
+            static::GET_MAIL_THEME_FOLDER_HOOK,
+            [
+                'mailTheme' => $mailTheme,
+                'mailThemeFolder' => &$mailThemeFolder,
+            ]
+        );
+        $this->checkThemeFolder($mailThemeFolder);
+
+        $mailThemeLayouts = new MailLayoutCollection();
+        $this->listCoreLayouts($mailThemeLayouts, $mailThemeFolder);
+        $this->listModulesLayouts($mailThemeLayouts, $mailThemeFolder);
+
+        //This hook allows to add/remove layouts to a mail theme
+        $this->hookDispatcher->dispatchWithParameters(
+            MailLayoutCatalogInterface::LIST_MAIL_THEME_LAYOUTS_HOOK,
+            [
+                'mailTheme' => $mailTheme,
+                'mailThemeLayouts' => $mailThemeLayouts,
+            ]
+        );
+
+        return $mailThemeLayouts;
+    }
+
+    /**
+     * @param MailLayoutCollectionInterface $collection
+     * @param string $mailThemeFolder
+     */
+    private function listCoreLayouts(MailLayoutCollectionInterface $collection, $mailThemeFolder)
+    {
+        $coreLayoutsFolder = implode(DIRECTORY_SEPARATOR, [
+            $mailThemeFolder,
+            MailTemplateInterface::CORE_CATEGORY,
+        ]);
+        if (!is_dir($coreLayoutsFolder)) {
+            return;
+        }
+
+        $this->addLayoutsFromFolder($collection, $coreLayoutsFolder);
+    }
+
+    /**
+     * @param MailLayoutCollectionInterface $collection
+     * @param string $mailThemeFolder
+     */
+    private function listModulesLayouts(MailLayoutCollectionInterface $collection, $mailThemeFolder)
+    {
+        $moduleLayoutsFolder = implode(DIRECTORY_SEPARATOR, [
+            $mailThemeFolder,
+            MailTemplateInterface::MODULES_CATEGORY,
+        ]);
+        if (!is_dir($moduleLayoutsFolder)) {
+            return;
+        }
+
+        $moduleFinder = new Finder();
+        $moduleFinder->directories()->in($moduleLayoutsFolder)->depth(0);
+
+        /* @var SplFileInfo $fileInfo */
+        foreach ($moduleFinder as $moduleFileInfo) {
+            $moduleName = $moduleFileInfo->getFilename();
+            $moduleFolder = implode(DIRECTORY_SEPARATOR, [$moduleLayoutsFolder, $moduleName]);
+            $this->addLayoutsFromFolder($collection, $moduleFolder, $moduleName);
+        }
+    }
+
+    /**
+     * @param MailLayoutCollectionInterface $collection
+     * @param string $folder
+     * @param string $moduleName
+     */
+    private function addLayoutsFromFolder(
+        MailLayoutCollectionInterface $collection,
+        $folder,
+        $moduleName = ''
+    ) {
+        $layoutFiles = [];
+        $finder = new Finder();
+        $finder->files()->in($folder)->sortByName();
+        /** @var SplFileInfo $fileInfo */
+        foreach ($finder as $fileInfo) {
+            //Get filename without any extension (ex: account.html.twig -> account)
+            $layoutName = preg_replace('/\..+/', '', $fileInfo->getBasename());
```

Comment:
What if you have `create.account.html.twig`? 
Not all users use `_` or `-` :(
Could be easier to remove known extensions 'html, txt, html.twig, wdyt?

Reply:
You're right but I don't want to encourage the use of `.` everywhere
As we already have a "double" extension with `.html.twig` I prefer the file base name to use `_` or `-` and not dotes.
This FolderCatalog is a very basic implementation of `MailLayoutCatalogInterface` so yes it is based on some conventions, but I don't want to complexify it more. It will likely evolve as themes definition will (maybe we will use a config file, or database persistence). So for now it's enough.

Besides (and maybe the more important), the template name doesn't bear any responsibility it is only a small description to identify the template so it's not that important.